### ISSUE TITLE
wallet builds invalid URL

### ISSUE THREAD

**Issue by bladedoyle** [mimblewimble/grin#2700]
**Describe the bug**
It looks like the wallet is building the url incorrectly when sending grin to http[s] address.  For the case where URL arguments are provided (poloniex) the ```/v1/wallet/foreign/receive_tx``` part of the url is being appended to the end of the argument list.

I guess poloniex attempted to "work around" this, but as a result the order of the arguments matters (which they should not).

**To Reproduce**
Attempt to send grin to poloniex with the URL arguments in a different order - ie, dont put the "id" argument last.

**Screenshots**
This works (id argument last):
```
20190321 18:20:58.846 INFO grin_wallet::command - Tx created: 1.000000000 grin to https://poloniex.com/public?currency=GRIN&command=createDeposit&id=8161xxxx-73xx-47xx-b9xx-f66e2cexxxxx (strategy 'all')
20190321 18:20:58.847 DEBUG grin_wallet::adapters::http - Posting transaction slate to https://poloniex.com/public?currency=GRIN&command=createDeposit&id=8161xxxx-73xx-47xx-b9xx-f66e2cexxxxx/v1/wallet/foreign/receive_tx
```

This does not work (id argument not last):
```
20190321 18:20:07.795 INFO grin_wallet::command - Tx created: 1.000000000 grin to https://poloniex.com/public?command=createDeposit&id=8161xxxx-73xx-47xx-b9xx-f66e2cexxxxx&currency=GRIN (strategy 'all')
20190321 18:20:07.795 DEBUG grin_wallet::adapters::http - Posting transaction slate to https://poloniex.com/public?command=createDeposit&id=8161xxxx-73xx-47xx-b9xx-f66e2cexxxxx&currency=GRIN/v1/wallet/foreign/receive_tx
```

**Comment by hashmap** [comment]
In theory we should fix it by appending `/v1/wallet/foreign/receive_tx` to the path part of URL, not the end of the string.

In practice using query string in `POST` request is quite rare and obviously Poloniex API is not a REST one, so using an opinionated REST client with it is problematic. Moreover there is a chance that `https://poloniex.com/public/v1/wallet/foreign/receive_tx?currency=GRIN&command=createDeposit&id=8161xxxx-73xx-47xx-b9xx-f66e2cexxxxx` will also fail.

So another option is to introduce a flag or smth like that to prevent URL append.

Last but not least option is to require 3rd party API to support Grin wallet URL scheme.

**Comment by rentenmark** [comment]
I sincerely wish this was not mishandled as it was but unfortunately this is what we are stuck with. The line in question is [`wallet/src/adapters/http.rs:48` ](https://github.com/mimblewimble/grin/blob/master/wallet/src/adapters/http.rs#L48) at time of writing 

```rust
let url = format!("{}/v1/wallet/foreign/receive_tx", dest);
```

This is obviously a naive approach and not appropriate for the more general web, unfortunately this is what we are all left with after 1.0 and what we have all implemented to. 

It is my opinion that this will probably have to stay as a sad implementation barnacle for the foreseeable future in order to maintain backward compatibility and avoid client/customer confusion. The current approach for anyone doing http sends is straightforward, if you are given a deposit url, simply deposit to that URL, nothing more, nothing less.

**Comment by hashmap** [comment]
Since this issue affects Poloniex and @rentenmark suggest to keep the existing behaviour I'm closing it. Feel free to reopen if needed

**Comment by bladedoyle** [comment]
I have no idea why this is close without any fix.

**Comment by hashmap** [comment]
I agree, however only Poloniex is affected by this issue and their developer suggested to not change this behavior. It seems we can make more harm by fixing it. If I got it wrong I’m fine to reopen and fix it.

**Comment by hashmap** [comment]
It seems the technical details were not mentioned and I don’t know how it was implemented on Poloniex side, but I can say that if we fix it they would need to support both behaviors, so I see disadvantages if a fix.

**Comment by bladedoyle** [comment]
As of right now, the only known exchange impacted is poloniex.  But leaving this bug will likely mean more implementations are impacted in the future.   The longer we wait, the more disruptive a fix will be.

Let me suggest that you schedule the fix for some future release, and notify Poloniex of the fix and the release number so they have sufficient opportunity to react.

**Comment by kargakis** [comment]
Can we ensure this gets fixed in the v2 api, if it's still a problem there?

**Comment by bladedoyle** [comment]
So far the concern is with how much impact this will have on one exchange, Poloniex.  Lets consider impact on others as well:

Any code that processes a payment URL that includes arguments will have to special-case each individual argument list.  My payment code, as an example, accepts a poloniex payment url via REST, and in order to make it work I have to do the following:

```
...
                    if "poloniex" in address: # Fuck, ugly - https://github.com/mimblewimble/grin/issues/2700
                        arg_data += "currency=" + request.json["currency"] + \
                                    "&command=" + request.json["command"] + \
                                    "&id=" + request.json["id"]
                    else:
                        for arg, val in request.json.items():
                            arg_data += arg + "=" + val + "&"
...
```

I'll need to add a new case for each exchange or web-based wallet that uses an argument list.  I'll have to be notified of the problem by users, debug the issue, add the special case, test that it didnt break anything else, and push the changes into production.

Every service / pool / wallet / etc, will need to go through this, not only me.  Please weigh that against the current impact of a single exchange needing to update due to a fix.   

One last point is, when Poloniex notices this bug (which they certainly must have) they could have:
1) reported the issue themselves
2) implemented the workaround in a more robust way

**Comment by hashmap** [comment]
Fair enough. Let’s target for 1.1 or so

**Comment by bladedoyle** [comment]
@hashmap   
I saw the comment by  kargakis but it didnt sink in at first.

Since there is an API V2 on its way, it seems fine to *not* fix in the v1 API - assuming the fix will be in the V2 api?

**Comment by kargakis** [comment]
cc @yeastplume
